typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1002 @@
+```
+
+
+#### Row polymorphism
+
+One more polymorphism is row.
+It allows you to make sure that the record contains at least given set of fields.
+Unfortunately TypeScript type compatibility is based on structural subtyping, so what we have available it is not _really_ a row polymorphism.
+Despite this we can use intersection types to emulate this.
+But you should be careful with using it because during assignments additional information can be lost.
+Developer must be cautious to propagate the full type-level information when they need it.
+
+```typescript
+type fnT = <T>(v: T & { x: number }) => T & { x: number }
+// You should fully describe all types to not lose information.
+
+type A = {};
+type B = A & {x: number};
+
+let a: A = {};
+const b: B = {x: 1};
+a = b;
+a.x; // Property 'x' does not exist on type 'A'.
+```
+
+### Mapped types, Conditional types and Type families
+
+One useful type functionality of TypeScript are Mapped and Conditional types.
+They give us really flexible ability to modify existing types or create now one.
+
+#### Mapped types
+
+Mapped types allows you to create new types based on old types transforming them in some way.
+The syntax looks like `[set of field names]: type;`.
+For example, you can add optional modifier or readonly modifier to each field with `type Partial<T> = { [P in keyof T]?: T[P] }` and `type Readonly<T> = { readonly [P in keyof T]: T[P] }`.
+It is also possible to remove such modifiers using `-?` and `-readonly`.
+
+The `keyof` operator in the code below produces a union of properties name.
+Square brackets indicates computed names.
+And `in` operator goes over all field keys.
+
+`Partial` type adds `?` modifier for each property.
+
+```typescript
+type Partial<T> = { [P in keyof T]?: T[P] };
+
+type A = { x: number; y: number };
+type PartialA = Partia<A>;
+
+const a: A = { x: 1 }; // Property 'y' is missing in type '{ x: number; }' but required in type 'A'.
+const a: PartialA = { x: 1 }; // Ok
+```
+
+Let's try now to remove `?` and `readonly` modifier from all properties.
+
+```typescript
+type NoModifiers<T> = { -readonly [P in keyof T]-?: T[P] };
+
+type A = { readonly x?: number; readonly y?: number };
+type NoModifiersA = NoModifiers<A>;
+
+const a: A = { x: 1 }; // Ok
+a.x = 1; // Cannot assign to 'x' because it is a read-only property.
+
+const a: NoModifiersA = { x: 1 }; // Property 'y' is missing in type '{ x: number; }' but required in type 'NoModifiers<A>'
+const a: NoModifiersA = { x: 1, y: 2 }; // Ok
+a.x = 2; // Ok
+```
+
+Let's present some complex examples of mapped types usage.
+
+Defining type which contains string and numerical properties.
+
+```typescript
+type Point = {
+  x: number;
+  y: number;
+  0: string;
+}
+```
+
+Now, using mapped types we will create a type which will pick only some properties.
+Word `extends` here say that generic `K` will contain a subset of union properties keys `T`.
```

Comment:
contain a subset of union properties keys `T` 

Something is not correct here grammar-wise, I can't parse it.

Reply:
> Will contain a subset of the properties of T.

Yes, it should work

> In particular, is there a specific reason why you need to mention union all the time?

Hm, now I understand that I mixed up `union` with `unit`. So, mentioning `unit` is like true name of what we have. I can mention it for example two times and then use just `set of properties` for example.

Reply:
Oh haha. :) Yeah, that should work.

Reply:
Grokked it and changed it.